web/modules/custom/weather/src/Plugin/Block/WeatherBlock.php

```diff
@@ -28,29 +31,56 @@ class WeatherBlock extends BlockBase {
    */
   public $method = "GET";
   /**
-   * For GuzzleHttp object.
+   * Guzzle\Client instance.
    *
-   * @var object
+   * @var \GuzzleHttp\ClientInterface
    */
   protected $client;
+  /**
+   * CacheBackendInterface instance.
+   *
+   * @var object
+   */
+  protected $cacheBackend;
+
+  /**
+   * {@inheritdoc}
+   */
+  public function __construct(
+    array $configuration,
+    $plugin_id,
+    $plugin_definition,
+    CacheBackendInterface $cache,
+    Client $http_client
+  ) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+    $this->client = $http_client;
+    $this->cacheBackend = $cache;
+  }
 
   /**
-   * Construct GuzzleHttp\Client object.
+   * {@inheritdoc}
    */
-  public function __construct($client) {
-    $this->client = new Client();
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('cache.default'),
+      $container->get('http_client'),
+    );
   }
 
   /**
    * {@inheritdoc}
    */
   public function build() {
-    $weather = $this->getWeather();
+    $cachedWeather = $this->getCachedWeather();
     return [
       '#theme' => 'weather',
-      '#name' => $weather["location"]["name"],
-      '#temperature' => $weather["current"]["temp_c"],
-      '#icon' => $weather["current"]["condition"]["icon"],
+      '#name' => $cachedWeather["location"]["name"],
+      '#temperature' => $cachedWeather["current"]["temp_c"],
+      '#icon' => $cachedWeather["current"]["condition"]["icon"],
```

Comment:
$cachedWeather can be NULL.
please use Null Coalescing Operator to avoid PHP notices
e.g. `'#name' => $cachedWeather["location"]["name"] ?? '',`